File: C_plus_plus/Search_Insert_Position.cpp
class Solution {
public:
    int searchInsert(int A[], int n, int target) {
        int i = 0 , j = n-1;
        while(i<=j){
            if(target==A[(i+j)/2]) return (i+j)/2;
            if(target<A[(i+j)/2]) j = (i+j)/2-1;
            if(target>A[(i+j)/2]) i = (i+j)/2+1;
        }
        return i>j?i:j;
    }
};

File: java/Search_Insert_Position.java

// Binary search
// we should be careful with the situation of the end.
// we use a flag to determine whether add one to the mid or keep mid to be our result.
// time is O(logn)
public class Solution {
    public int searchInsert(int[] nums, int target) {
        int low = 0, high = nums.length - 1;
        int flag = 0, mid = 0;
        while (low <= high) {
            mid = (low+high)/2;
            if (nums[mid] == target) return mid;
            else if (nums[mid] > target) {
                high = mid - 1;
                flag = 0;
            }
            else {
                low = mid + 1;
                flag = 1;
            }
        }
        return mid + flag;
    }
}
